### PR TITLE
Friendship v2

### PR DESCRIPTION
**Pull request by wathigo**
Tasks Description
* Made User model methods more efficient.
* Clean up code

### CODE REVIEW

**Comment by wathigo** [comment]
**Thank you @adriaanbd  for the review :)**
You are right. I failed to update my helper methods after updating my code and this could be avoided if my app was well tested.

**Comment by wathigo** [comment]
I will fix the bug and also add some tests

**Comment by wathigo** [comment]
For the routes, I will not change them since I used custom names so that I do not confuse them while using them in other parts of the application.

File: fb/app/models/user.rb [hunk 1]
```diff
@@ -57,33 +46,34 @@ def friends?(user)
...
+      ((friends + unknown_users(self)) - [self]) - pending_requests
     else
-      (friends - [self]) - inverse_pending_friends
+      (friends - [self]) - pending_requests
     end
   end
 
   def confirmed_friends
-    friendships.map { |friendship| friendship.friend if friendship.confirmed }.compact
+    confirmed_friendships.map { |friendship| friendship.friend }
```
Comment (mosaaleb) [inline fb/app/models/user.rb]: - [ ] We can use instead use `has_many :confirmed_friends, through: :confirmed_friendships`. This association will also give you a number of helper method like `confirmed_friend_ids` and many other. Please take a look at [this](https://api.rubyonrails.org/classes/ActiveRecord/Associations/ClassMethods.html) under "collection associations" section to check other generated methods from this association.

File: fb/app/models/user.rb [hunk 2]
```diff
@@ -57,33 +46,34 @@ def friends?(user)
...
   end
 
   def confirmed_inverse_friends
-    inverse_friendships.map { |friendship| friendship.user if friendship.confirmed }.compact
+    confirmed_inverse_friendships.map { |friendship| friendship.user }
```
Comment (mosaaleb) [inline fb/app/models/user.rb]: - [ ] Similarly we can use `has_many :inverse_friends, through: :confirmed_inverse_friendships`.

File: fb/app/models/user.rb [hunk 3]
```diff
@@ -32,20 +35,6 @@ def mutual_friends(user)
...
-  def pending_inverse_friends
-    friends2 = inverse_friendships.map { |friendship| friendship.user unless friendship.confirmed }.compact
-  end
-
-  def friend_requests
-    inverse_friendships.map { |friendship| friendship.user unless friendship.confirmed }.compact
-  end
-
   def confirm_friend(requester)
     friendship = inverse_friendships.find { |friendship| friendship.user == requester }
```
Comment (mosaaleb) [inline fb/app/models/user.rb]: - [ ] when friendship is confirmed, an additional row with reversed user-friend pair is created.

Comment (wathigo) [reply fb/app/models/user.rb]: Thank you. I will fix this too.

File: fb/app/models/user.rb (hunk 2 above)
Comment (wathigo) [reply fb/app/models/user.rb]: Thank you for providing the resource, it was very helpful

File: fb/app/models/user.rb (hunk 1 above)
Comment (wathigo) [reply fb/app/models/user.rb]: Thank you for providing the resource, it was very helpful

**Comment by wathigo** [comment]
Thank you for the review. @Oghenebrume50 . I included some posts if a user does not have friends or any posts because of the structure of the app. The same effect when there are no requests is the same as the one when there are no posts. I will however make the necessary changes.

**Comment by wathigo** [comment]
Thank you @mcrd25  for catching this. I have fixed it

**Pull request merged**